### PR TITLE
Package update

### PR DESCRIPTION
**Pull request by Andreas-Dorfer**
Hi @shuebner ,

I really like your project. It makes using OneOf feel way more natural in C#. Unfortunately, one of the package references indirectly points to a package with a known vulnerability (see [System.Net.Http](https://www.nuget.org/packages/System.Net.Http/4.3.0)):

`SvSoft.OneOf.Analyzers.SwitchDiagnosticSuppression@1.1.0 › Microsoft.CodeAnalysis.CSharp.Workspaces@3.11.0 › Humanizer.Core@2.2.0 › NETStandard.Library@1.6.1 › System.Net.Http@4.3.0`

It's not an actual issue, since your diagnostics suppressor is only a build-time dependency, but it can get flagged by package vulnerability scans that don't recognize it as build-time dependency - which can be annoying.

With this PR, I
- update all package references
- update the SDK version to `8.0.300`
- update the test project to `net8.0`
- update the language version to `12`
- apply code suggestions
- update the GitHub actions to `@v4`

There is one necessary fix in `DiagnosticSuppressorAnalyer.cs` because of a change in behavior of `GetAllDiagnosticsAsync` (probably caused by dotnet/roslyn#53045 ). To get the same result as with the currently used version of `Microsoft.CodeAnalysis.CSharp.Workspaces`, I have to explicitly create `CompilationWithAnalyzersOptions` where `reportSuppressedDiagnostics` is set to `true`.

If you dislike any of the updates I did, let me know and I will change the PR accordingly.

Thank you,
Andreas

### CODE REVIEW

**Comment by shuebner** [comment]
Hi @Andreas-Dorfer ,
thank you for the kind words. And thank you for contributing. As you can see from the commit and issue history, I invest next to no time here.
I am on vacation right now. I will try to review the PR in a week. Judging from what you wrote in the description, you really thought this through, so I expect a quick merge here.

I am also happy to see that someone actually uses this library. And even happier that this issue is not even a bug report ;-)

File: OneOfDiagnosticSuppressor/OneOfDiagnosticSuppressor.csproj [hunk 1]
```diff
@@ -23,7 +24,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.11.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="4.9.2" />
```
Comment (shuebner) [inline OneOfDiagnosticSuppressor/OneOfDiagnosticSuppressor.csproj]: This breaks compatibility with all SDKs but the latest one.
Please use the lowest version without the problematic Http reference.

**Review by shuebner — CHANGES_REQUESTED** [review]
I added a commit to main adding PrivateAssets="all" to the CodeAnalysis package. It seems I forgot to fix that. In my other package (ClosedTypeHierarchyDiagnosticSuppressor) I fixed that a long time ago, but apparently forgot to to the same here.

It may well be that this PrivateAssets is already the solution to the dependency analysis warnings.

In any case, the vulnerability check you use seems to not take into account the actual dependency tree. As long as you target at least netstandard2.0, the Humanizer package does not depend on the net standard lib 1.6 with the problematic Http package.
That dependency only exists when nuget resolves netstandard1.0 as the best target.
Which tool are you using?

I will probably not support this analyzer for that low a target anyway.

File: OneOfDiagnosticSuppressor/OneOfDiagnosticSuppressor.csproj (hunk 1 above)
Comment (Andreas-Dorfer) [reply OneOfDiagnosticSuppressor/OneOfDiagnosticSuppressor.csproj]: changed version of `Microsoft.CodeAnalysis.CSharp.Workspaces` to `4.3.1`

**Comment by Andreas-Dorfer** [comment]
Hi @shuebner , thanks for looking at my PR!

Adding PrivateAssets="all" might actually do the trick. It could be what's missing to recognize it as build-time dependency. The tool I use is [Snyk](https://snyk.io/product/open-source-security-management/).

About the dependencies: I might be missing something, but when I look at `Humanizer.Core`'s dependencies for version `2.2.0` on [nuget](https://www.nuget.org/packages/Humanizer.Core/2.2.0#dependencies-body-tab), it does use `NETStandard.Library` `1.6.1`. 
It's only starting from `Humanizer.Core` version `2.3.2`, that it no longer depends on `NETStandard.Library` `1.6.1` when targeting `.NETStandard 2.0`.

**Comment by shuebner** [comment]
You are of course right about Humanizer.Core. I must have accidentally checked a different version of Humanizer.Core.

If I understood you correctly, my fix of declaring PrivateAssets on the Microsoft.CodeAnalysis.CSharp.Workspaces package within the analyzer (as it should have been from the beginning) already solved your problem.

If that is true, this PR is no longer necessary.
I would publish a patch from the current master commit instead.

BTW, the PR in its current form would greatly restrict compatibility, because it introduces a dependency on the latest version of the Microsoft.CodeAnalysis.CSharp _assembly_ at dotnet compile time.
This would then only work with the latest dotnet SDK and the latest Visual Studio.
Older versions would get an assembly load exception from the analyzer at compile time.

If not strictly necessary, I would not touch the Microsoft.CodeAnalysis.CSharp package version for now to keep compatibility with old SDKs.

**Comment by shuebner** [comment]
@Andreas-Dorfer 
Since I have not heard from you, I have published a package version 1.1.1 with the fixed dependencies.
If this fixes your problem, I will close this PR.

In any case, thank you for drawing my attention to this issue :-)

**Comment by Andreas-Dorfer** [comment]
I'll do a Snyk scan on 1.1.1 and let you know the result.

**Comment by Andreas-Dorfer** [comment]
v 1.1.1 doesn't get flagged by Snyk. Thanks @shuebner !

**Pull request closed without merging**